stl/inc/ostream

```diff
@@ -1076,6 +1082,265 @@ basic_ostream<_Elem, _Traits>& operator<<(basic_ostream<_Elem, _Traits>& _Ostr,
+    const ostream::sentry _Ok(_Ostr);
+    ios_base::iostate _State = ios_base::goodbit;
+
+    if (!_Ok) [[unlikely]] {
+        _State |= ios_base::badbit;
+    } else [[likely]] {
+        if constexpr (_Add_newline) {
+            // This is intentionally kept outside of the try/catch block in _Print_noformat_nonunicode_impl()
+            // (see [ostream.formatted.print/3.2]).
+            const string _Output_str{string{_Str} + '\n'};
+            _State |= _STD _Print_noformat_nonunicode_impl(_Ostr, _Output_str);
+        } else {
+            _State |= _STD _Print_noformat_nonunicode_impl(_Ostr, _Str);
+        }
+    }
+
+    _Ostr.setstate(_State);
+}
+
+template <bool _Add_newline, class... _Types>
+void _Print_impl(ostream& _Ostr, const _Print_string<_Types...> _Fmt, _Types&&... _Args) {
+    // We can expect the case of using std::print() or std::println() without any format
+    // strings to be common enough to warrant optimization. We can actually check at compile
+    // time as to whether or not any formatting needs to be done by checking if both the
+    // string does not have any escaped brace characters (i.e., {{ or }}) and
+    // sizeof...(_Types) == 0; in that case, we don't need to use std::format. This should
+    // lead to much better code generation.
+    constexpr bool _Requires_formatting = (sizeof...(_Types) > 0);
+
+    if constexpr (_Is_ordinary_literal_encoding_utf8) {
+        if constexpr (_Requires_formatting) {
+            _STD _Vprint_unicode_impl<_Add_newline>(
+                _Ostr, _Fmt._Get(), _STD make_format_args(_STD forward<_Types>(_Args)...));
+        } else {
+            const typename _Print_string<_Types...>::_Format_string_type _Str_type = _Fmt._Get_type();
+            if (_Str_type == _Print_string<_Types...>::_Format_string_type::_Requires_formatting) {
+                _STD _Vprint_unicode_impl<_Add_newline>(
+                    _Ostr, _Fmt._Get(), _STD make_format_args(_STD forward<_Types>(_Args)...));
+            } else {
+                _STD _Print_noformat_unicode<_Add_newline>(_Ostr, _Fmt._Get());
+            }
+        }
+    } else {
+        if constexpr (_Requires_formatting) {
+            _STD _Vprint_nonunicode_impl<_Add_newline>(
+                _Ostr, _Fmt._Get(), _STD make_format_args(_STD forward<_Types>(_Args)...));
+        } else {
+            const typename _Print_string<_Types...>::_Format_string_type _Str_type = _Fmt._Get_type();
+            if (_Str_type == _Print_string<_Types...>::_Format_string_type::_Requires_formatting) {
+                _STD _Vprint_nonunicode_impl<_Add_newline>(
+                    _Ostr, _Fmt._Get(), _STD make_format_args(_STD forward<_Types>(_Args)...));
+            } else {
+                _STD _Print_noformat_nonunicode<_Add_newline>(_Ostr, _Fmt._Get());
+            }
+        }
+    }
+}
+
+_EXPORT_STD template <class... _Types>
+void print(ostream& _Ostr, const _Print_string<_Types...> _Fmt, _Types&&... _Args) {

```

Comment:
(Standardized) `format_string` is currently used in the specification ([[ostream.formatted.print]](https://eel.is/c++draft/ostream.formatted.print), [[print.fun]](https://eel.is/c++draft/print.fun)). I think it's required to exactly use it for standard functions, and thus `_Print_string` should only be used in implementation details (`_Print_impl` etc.).

For example, given the following type
```C++
struct FmtSrc {
    template<std::same_as<std::format_string<int>> T>
    constexpr operator T() const noexcept {
        return "{}";
    }
};
```
`std::print(std::cout, FmtSrc{}, 42)` (and similar function calls) is required to be well-formed and behave same as `std::print(std::cout, "{}", 42)` IIUC.

Ditto for other `print` and `println` overloads.

Reply:
Yeah, I wasn't sure if changing standard function signatures like that would be allowed, but I assumed they would be under the "as-if" rule, so long as situations like the one you just mentioned don't interfere with the ability to call the function. Here's something that seems to work for your example:

```C++
template <class _CharT, class... _Types>
struct _Basic_print_string : private basic_format_string<_CharT, _Types...> {
public:
    template <class _Ty>
        requires convertible_to<const _Ty&, basic_string_view<_CharT>>
    consteval _Basic_print_string(const _Ty& _Str_val) : basic_format_string<_CharT, _Types...>(_Str_val) {
        _Initialize_format_string_type();
    }

    template <class _Ty>
        requires (convertible_to<const _Ty&, basic_format_string<_CharT, _Types...>>
                   && !convertible_to<const _Ty&, basic_string_view<_CharT>>)
    constexpr _Basic_print_string(const _Ty& _Str_val) : basic_format_string<_CharT, _Types...>(_Str_val) {
        _Initialize_format_string_type();
    }

    // ...
};
```

The big reason for using the `_Print_string` type was to use a `consteval` constructor to check for escaped brace characters at compile time. I don't want to abandon the idea behind this optimization completely, but your example is definitely cause for concern. I'll have to look into this more on my own (whether function signature transformations like this are even allowed, what needs to be done to account for edge cases, etc.). For now, though, I'll probably be removing `_Print_string` and the associated optimization for the time being.

Reply:
Actually, there is another way: We could extend `std::basic_format_string` to do the escaped braces check as part of its `consteval` constructor. Then, we could have some private/internal member function of that type return whether or not the string is directly printable! This solves the problem of needing function signatures to match while still allowing us to implement the original optimization. (Why didn't I think of that before...?)

Reply:
After further investigation, it looks like the C++ standard explicitly disallows modifying the function signatures of non-member STL functions (see [[global.functions/2]](https://eel.is/c++draft/library#global.functions-2), along with the associated [footnote](https://eel.is/c++draft/library#footnote-167)). So, the whole idea was basically a wash from the start; even if we did account for every edge case, the implementation wouldn't comply with the standard. It's frustrating, but we should always prioritize conformance.

What this means in practice is that we can't check at compile time whether escaped braces are present in the user's string. The current implementation of `_Print_impl()` now manually copies format strings while accounting for escaped braces as necessary when no formatting arguments are present. This should still be faster than a direct call to `std::format()`.